Assign quizzes to individual learners

nucleogenesis opened the pull request:
<!--
 1. Following guidance below, replace …'s with your own words
 2. After saving the PR, tick of completed checklist items
 3. Skip checklist items that are not applicable or not necessary
 4. Delete instruction/comment blocks
-->

### Summary
<!--
 * description of the change
 * manual verification steps performed
 * screenshots if the PR affects the UI
-->

Implements the ability to assign individual learners per the requirement listed in https://www.notion.so/learningequality/UX-UI-design-499b292122a142a1b23124d38a76c420

One notable difference between the mock-up (https://www.figma.com/proto/TcSczscKh4zLTypFJh2Or4/0.13-Lesson-%26-quiz-updates?node-id=176%3A151&scaling=scale-down) and the implementation is that there is no "User type" filter - every user in this context is a Learner so it doesn't need to be there.

### Reviewer guidance
<!--
 * how can a reviewer test these changes?
 * are there any risky areas that deserve extra testing
-->

_Please test this using a previously used KOLIBRI_HOME folder. Copy it somewhere and delete it after testing. This PR requires that a new data model be associated with quizzes. When you create a NEW quiz - the logic handles it there. However, there is additional logic that ensures that previously created quizzes (which were created prior to the logic that now exists to include the association upon quiz creation) will also have the appropriate association created._

__Step 1: Assign an Individual Learner__

- Go to Coach > Class > Plan > Quizzes > Quiz Details for a quiz created before running this PR.
- Ensure that the Quiz is "Open"
- Go to the Options > Edit Details
- Uncheck any previously checked Groups if there are any.
- See that the "Individual learners" checkbox exists.
- Click it and see that the Learners assigned to this class are listed there.
- Click one or more of the checkboxes to select a Learner - remember the username of that Learner.
- While the Individual Learners table is open - click a group and see that the learners in the table are disabled if they are part of the selected group.
- Test the filter input for the table please!
- Save

__Step 2: See that the Learner is Assigned__

- In a separate browser window (Incognito / Private) login as the Learner to whom the quiz was assigned.
- See that the learner may take the quiz.
- See that in the Quiz Status sidebar (In Quiz Details) that the Group and Learners names are listed under "Recipients" - if you have 3 other groups selected, you won't see learners' names but you'll see "Group 1, Group2 and N others" - but if you only have one group and some individual learners selected, it should read like "Group 1, user 1 and N others". If it is ONLY individual learners selected, it should be names of learners.

__Step 3: Try to break assigning learners__

- Return to the browser where you are logged in as the Coach.
- Return to the "Edit Details" page for the same quiz.
- Now, select a Group. If one doesn't exist, go back and create a group and assign some but not all learners to it.
- See that the table of Individual learners is now filtered so that the members of the selected group are not visible.
- Now - select several learners individually or as a part of a group and verify that they are able to see the quiz (or not be able to see it) as expected.
- Do the same for several other quizzes as well.

__Step 4: Create a new quiz__

- Go to Coach > Class > Plan > Quizzes > New Quiz
- Create a new quiz
- Replay "Step 1" and "Step 2" for the newly created quiz. 


__Step 5: Side Effects__

- Create a quiz and assign it to a group - then assign it to a number of individual learners. Remember the number of individual learners.
- Go through everything in the Quiz Reports - take quizzes as Individual Learners and not Invidiual learners - are the number reported properly?
- Ensure that the Class Home is reflective of the IndividualLearners and their behavior as well as that of the learners in normal groups.

__Step 5.5: Side Effects (Again)__

- Do Step 5, but ONLY assign individual learners - no other groups.
- Make sure everything works as expected.

### References
<!--
 * references to related issues and PRs
 * links to mockups or specs for new features
 * links to the diffs for any dependency updates, e.g. in iceqube or the perseus plugin
-->

https://www.figma.com/proto/TcSczscKh4zLTypFJh2Or4/0.13-Lesson-%26-quiz-updates?node-id=176%3A151&scaling=scale-down

https://github.com/learningequality/clearinghouse/issues/422

https://github.com/learningequality/clearinghouse/issues/352

----

### Contributor Checklist


PR process:

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

Testing:

- [x] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

### Reviewer Checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

nucleogenesis committed:
Assign individual learners

- Add new collection kind
- Wire API to pass IndividualLearner collection data to Vuex
- Build selection table and some logic
- Add several $trs

nucleogenesis committed:
Configuring Individual Selector & Lint & Strings

nucleogenesis committed:
Renaming migration file for individual learners

nucleogenesis committed:
Add itemsPerPage prop to PaginatedListContainer

- itemsPerPage prop now allows a Number to be passed to customize the
quantity of items on each page of the list.
- This also allows the parent component to use that number in its own
computed properties

- Additionally, it now emits a 'pageChanged' event whenever the page number
changes.

- Also, instead of watching `filteredItems` - which could "change" even if
there is no meaningful change to the listed items,
this will watch the `numFilteredItems` - forcing the table to page 1
whenever the number of filtered items changes - which is really when it
matters to the user.

nucleogenesis committed:
Backend IndividualLearnersGroup

- Add top-level comment on IndividualLearnersGroup class definition
- Add `get_learners()` to retrun learners associated to ILG

- Handle `partial_update()` in ILGViewSet to ensure learners selected on
frontend are properly associated with the ILG

- Remove `UniqueTogetherValidator` that I copied early on - the `name` of
ILGs are always the same (and useless) so we cannot have them checked
as unique in association with their parents because a class
will have several ILGs all of which are named "Individual learners"

- Exclude ILG collection IDs from `exam['groups']` and
`lesson['groups']` in ClassSummary

nucleogenesis committed:
Create and assign ILG on exam creation

- Create an ILG before submitting the exam
creation form
- Pass the ILGs ID to the createExamAndRoute process
- Include the ILG in assignments while creating the Exam

nucleogenesis committed:
Finalize ILG vuex actions and state

nucleogenesis committed:
Make ILG work in QuizEditDetailsPage et al

- Adds two new strings (oops...)
- Ensure data is passed by events between <ILG> and parents
- Table's selection of learners across pages (including
"entire page") works as expected
- Add tooltip and "hiding selected groups" functionality

- **Importantly** - if we get to the QuizEditDetailsPage and there is no ILG
assigned to the quiz, then we will create one and assign it.

nucleogenesis committed:
Assign ILG when created in QuizEditDetailsPage

nucleogenesis commented on kolibri/plugins/coach/assets/src/views/plan/assignments/IndividualLearnerSelector.vue:
```diff
@@ -186,6 +246,17 @@
+        message: 'Not showing learners selected from',
+        context:
+          'A label in a tooltip that explains which groups of learners are not being displayed in the table used to select individual learners.',
+      },
```
@indirectlylit @radinamatic 

I failed to include these in my earlier PR that added the other strings in this file. How should we handle this?

nucleogenesis replied:
These have been removed

nucleogenesis commented on kolibri/core/auth/api.py:
```diff
@@ -404,6 +406,42 @@ def annotate_queryset(self, queryset):
+
+    field_map = {"user_ids": partial(process_uuid_aggregate, key="user_ids")}
+
+    def annotate_queryset(self, queryset):
```
@rtibbles I've included this because it appeared in the LearnerGroupViewSet while I was rebasing - please advise if this is unnecessary here or if there is anything that needs to be changed.

Note that the InvididualLearnersGroup is also a child of Collection - so it seems that this code ought to work as expected assuming it is expected to be here in the first place.

nucleogenesis committed:
Black formatting...

indirectlylit commented:
In the design, we hide individual learners when they've already been implicitly selected by virtue of being in a selected group. Unfortunately we missed a couple strings related to this functionality (ref #6150).

## Proposed update

For sake of argument, let's say we have a class with only one learner, and that learner is assigned to Group 1. Then, learners who assigned via the group are...

| old behavior | proposed behavior |
|--|--|
| hidden from the individual learner table | shown, but with a disabled and checked box |
| ![image](https://user-images.githubusercontent.com/2367265/69681615-75a86b80-1063-11ea-9c4a-7d6a3968c63b.png) | ![image](https://user-images.githubusercontent.com/2367265/69681737-df287a00-1063-11ea-9dbe-c287ac095911.png)|

I think this actually might be more clear than the original specs and will cause less distracting shifting of the table rows.

nucleogenesis committed:
Change hidden learners styling +1

+1 is that I removed unused `mapState` from
QuizEditDetailsPage

nucleogenesis committed:
Use prop to determine quiz vs lesson, test data

- Added test data for ILG to ClassSummary fixtures

- Use `assignmentType` prop - pass to RecipientSelector to determine
whether we show the ILS component in the list of groups

- Found a regression in Lesson Edit Details. They both used the
`submitData` method but the Quiz has logic to manage
the individual learners data so I split them into
two methods `submitQuizData` and `submitLessonData`.

nucleogenesis committed:
Gherkin scenario preliminary

radinamatic commented on integration_testing/features/coach/coach-edit-quiz.feature:
```diff
@@ -4,6 +4,7 @@ Feature: Coach edits quizzes
   Background:
     Given I am signed in to Kolibri as a coach user
       And there is a quiz <quiz> created previously
+      And the quiz is started
```
I don't think this line is necessary, coach can make these changes even if the quiz in not started.

nucleogenesis replied:
This is because later on, I instruct the tester to verify that the learner can indeed take the quiz and the learner can only see the quiz if it is started / active.

radinamatic commented on integration_testing/features/coach/coach-edit-quiz.feature:
```diff
@@ -28,6 +29,17 @@ Feature: Coach edits quizzes
+    When I change *Recipients* by selecting *Individual learners*
+    Then I see a table listing all of the learners in the class.
+    When I change *Recipients* by selecting both *Individual learners* and any other group that has learners
+    Then I see the learners in that group have the checkboxes by their names disabled
```
`...learners in the selected group(s) have checkboxes...`

radinamatic commented:
Running this branch to test the Gherkin, but I notice that the individual learners when assigned alongside a group, are apparently not listed as recipients 😕 

![2019-11-27_23-48-42](https://user-images.githubusercontent.com/1457929/69764610-3ff1a880-1171-11ea-866c-8af17b3dfd1c.png)

User `abbyl` has been assigned the quiz as an individual learner, alongside the group B, and has started the quiz, but her activity is not being recorded.

![2019-11-27_23-47-27](https://user-images.githubusercontent.com/1457929/69764668-78918200-1171-11ea-85a7-40042ea49ec6.png)

No amount of refreshing was able to display that coach-side.

![2019-11-27_23-46-34](https://user-images.githubusercontent.com/1457929/69764700-965ee700-1171-11ea-8aa2-3c8f21a00bdc.png)

[logs.zip](https://github.com/learningequality/kolibri/files/3899091/logs.zip)

indirectlylit commented:
>  but I notice that the individual learners when assigned alongside a group, are apparently not listed as recipients

I think that is a known issue - @nucleogenesis flagged it here:

https://learningequality.slack.com/archives/C0LKG14NL/p1574887231248800

> User abbyl has been assigned the quiz as an individual learner, alongside the group B, and has started the quiz, but her activity is not being recorded.

this sounds like a bug

nucleogenesis commented:
@radinamatic @indirectlylit RE: The activity not showing up on the Class Home - I found the error and will be working on it - will have a fix pushed Monday most likely.

indirectlylit commented on kolibri/core/auth/constants/collection_kinds.py:
```diff
@@ -8,10 +8,12 @@
 FACILITY = "facility"
 CLASSROOM = "classroom"
 LEARNERGROUP = "learnergroup"
+INDIVIDUALLEARNERSGROUP = "individuallearnersgroup"
```
Is the function of this new collection kind:

1. A special "collection" that really should always have exactly 1 user in it
2. A new "collection" that is not shown to coaches, but represents an ad-hoc group of individual learners within a class?

nucleogenesis replied:
This collection creates a new kind of group.

- It is a collection that should always be assigned to one exam.
- It can have any number of learners in it who are also associated with the parent collection (classroom).
- It can be shown at will on the front-end. Currently, it is only shown to Coaches while in the Quiz > Edit Details page but it could relatively easily be included into the Lessons if desired down the road.
- It is not shown as a "group" in the Groups page at all, though if that's what you mean by "not shown to coaches"

indirectlylit commented on kolibri/core/auth/models.py:
```diff
@@ -1406,3 +1414,52 @@ def remove_learner(self, user):
+    """
+    An ``IndividualLearnersGroup`` is a collection kind that can be used in an assignment
+    to create a group that is specific to a single ``Lesson`` or ``Exam``.
+    """
```
as far as naming ... `IndividualLearnersGroup` and `LearnerGroup` are pretty similar.

Maybe a name like `LearnerSelection` or `AdHocGroup`? These imply to me that it's a one-off creation intended for things like assignments...

nucleogenesis replied:
Renamed throughout to AdHocGroup.

The only thing that didn't change is the `IndividualLearnerSelecter.vue` file because strings are associated with it.

indirectlylit commented on kolibri/core/content/test/test_import_export.py:
```diff
@@ -179,7 +179,7 @@ def test_remote_import_readtimeout(
 
 @patch(
-    "kolibri.core.content.management.commands.importcontent.calculate_files_to_transfer",
+    "kolibri.core.content.management.commands.importcontent.calculate_files_to_transfer"
```
not a big deal, but strange that this was necessary ... are you running the latest version of Black? I know we had some annoying version inconsistency issues recently

indirectlylit commented on kolibri/plugins/coach/assets/src/views/plan/assignments/IndividualLearnerSelector.vue:
```diff
@@ -91,44 +112,72 @@
-        showUserTable: false,
-        selectedIndividualIds: [],
-        selectAllOnPage: false,
+        isChecked: !!this.initialIndividualLearners.length,
```
```suggestion
        isChecked: Boolean(this.initialIndividualLearners.length),
```

Back in #3224 we standardized on this - general philosophy is that being explicit and improving readability is better than saving a few characters

nucleogenesis committed:
Add ILG getters to classSummary and commonCoach

- invidiualLearnersGroups getter gets the groups from the ilgMap

- getIndividualLearnersForExam getter gets the learners individually assigned

- getLearnersForExam is updated to return ILG learners as well

nucleogenesis committed:
Rename IndividualLearnersGroup to AdHocGroup

- Also do this everywhere else.

- IndividualLearnerSelector.vue has strings named for itself so we cannot
rename it right now.

nucleogenesis committed:
Ensure individual learners names in Recipients

- Class Home - ItemProgressCard - see names with groups

- QuizStatus - in quiz details pages, see names with groups

- Type casting explicitly with Boolean() rather than !!

nucleogenesis committed:
Pass the tests

- Add sampleState2 data for adHocLearnersMap

- Fix error I created that failed a test. Yay for tests!

nucleogenesis commented:
@radinamatic 

The PR is now updated so that the following issues are fixed:

- Class Home summary - now when individually assigned learners progress, their data is reflected properly here (and throughout the reports, which it wasn't previously)

- Recipients now will list learner names after groups (if there are groups selected).

**Note** that I have updated the Reviewer Guidance - let me know if you have any questions

Also - please note my response here: https://github.com/learningequality/kolibri/pull/6149#discussion_r351532481

nucleogenesis committed:
Gherkin wording update

indirectlylit commented:
@nucleogenesis I'm getting some critical errors on the report page:

![image](https://user-images.githubusercontent.com/2367265/70253408-02c08400-1738-11ea-9ef7-bb83a1f90fc6.png)

nucleogenesis committed:
Also use gropuAndAdHocLearnerNames prop in report

jonboiser committed:
Merge branch 'release-v0.13.x' into assign-individual-learners

radinamatic commented:
@indirectlylit Documenting this should have the 0.13.2 milestone, because of the upcoming translated user strings, correct?

indirectlylit commented:
0.14.0 now, because we've decided to push it to `develop` instead
